mm2src/mm2_main/src/ext_api/one_inch/errors.rs

```diff
@@ -0,0 +1,83 @@
+use coins::{eth::u256_to_big_decimal, NumConversError};
+use common::{HttpStatusCode, StatusCode};
+use enum_derives::EnumFromStringify;
+use mm2_number::BigDecimal;
+use ser_error_derive::SerializeErrorType;
+use serde::Serialize;
+use trading_api::one_inch_api::errors::ApiClientError;
+
+#[derive(Debug, Display, Serialize, SerializeErrorType, EnumFromStringify)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum ApiIntegrationRpcError {
+    #[from_stringify("coins::CoinFindError")]
+    CoinFindError(String),
+    #[display(fmt = "EVM token needed")]
+    CoinTypeError,
+    #[display(fmt = "NFT not supported")]
+    NftNotSupported,
+    #[display(fmt = "Chain not supported")]
+    ChainNotSupported,
+    #[from_stringify("coins::UnexpectedDerivationMethod")]
+    MyAddressError(String),
+    InvalidParam(String),
+    #[display(fmt = "allowance not enough for 1inch contract, available: {allowance}, needed: {amount}")]
+    OneInchAllowanceNotEnough {
+        allowance: BigDecimal,
+        amount: BigDecimal,
+    },
+    #[display(fmt = "1inch API error: {}", _0)]
+    OneInchError(ApiClientError),
+    ApiDataError(String),
+}
+
+impl HttpStatusCode for ApiIntegrationRpcError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            ApiIntegrationRpcError::CoinFindError(_)
+            | ApiIntegrationRpcError::CoinTypeError
+            | ApiIntegrationRpcError::NftNotSupported
+            | ApiIntegrationRpcError::ChainNotSupported
+            | ApiIntegrationRpcError::MyAddressError(_)
+            | ApiIntegrationRpcError::InvalidParam(_)
+            | ApiIntegrationRpcError::OneInchAllowanceNotEnough { .. } => StatusCode::BAD_REQUEST,
+            ApiIntegrationRpcError::OneInchError(_) | ApiIntegrationRpcError::ApiDataError(_) => {
+                StatusCode::BAD_GATEWAY
+            },
+        }
+    }
+}
+
+impl ApiIntegrationRpcError {
+    pub(crate) fn from_api_error(error: ApiClientError, decimals: u8) -> Self {
+        match error {
+            ApiClientError::InvalidParam(error) => ApiIntegrationRpcError::InvalidParam(error),
+            ApiClientError::HttpClientError(_)
+            | ApiClientError::ParseBodyError(_)
+            | ApiClientError::GeneralApiError(_) => ApiIntegrationRpcError::OneInchError(error),
+            ApiClientError::AllowanceNotEnough(nested_err) => ApiIntegrationRpcError::OneInchAllowanceNotEnough {
+                allowance: u256_to_big_decimal(nested_err.allowance, decimals).unwrap_or_default(),
+                amount: u256_to_big_decimal(nested_err.amount, decimals).unwrap_or_default(),
+            },
+        }
+    }
+}
+
+/// Error aggregator for errors of conversion of api returned values
+#[derive(Debug, Display, Serialize)]
+pub(crate) struct FromApiValueError(String);
```

Comment:
Is there any benefit of wrapping `String` with a struct instead of doing `type FromApiValueError = String`?

Reply:
I think we mostly use 'typed' errors (for easier processing and conversion), so should reduce errors as general types like String.